x-pack/plugins/security/server/anonymous_access/anonymous_access_service.ts

```diff
@@ -0,0 +1,140 @@
+    request: KibanaRequest,
+    savedObjectType: string
+  ) => Promise<boolean>;
+}
+
+interface AnonymousAccessServiceStartParams {
+  authz: Pick<AuthorizationServiceSetup, 'actions' | 'checkSavedObjectsPrivilegesWithRequest'>;
+  basePath: IBasePath;
+  spaces?: SpacesServiceStart;
+}
+
+/**
+ * Service that manages various aspects of the anonymous access.
+ */
+export class AnonymousAccessService {
+  constructor(private readonly logger: Logger, private readonly getConfig: () => ConfigType) {}
+
+  setup(): AnonymousAccessServiceSetup {
+    return {
+      isAnonymousAccessEnabled: this.isAnonymousAccessEnabled(),
+    };
+  }
+
+  start({
+    authz,
+    basePath,
+    spaces,
+  }: AnonymousAccessServiceStartParams): AnonymousAccessServiceStart {
+    const config = this.getConfig();
+    const anonymousProvider = config.authc.sortedProviders.find(
+      ({ type }) => type === AnonymousAuthenticationProvider.type
+    );
+    // We don't need to add any special parameters to the URL if any of the following is true:
+    // * anonymous provider isn't enabled
+    // * anonymous provider is enabled, but it's a default authentication mechanism
+    const anonymousIsDefault =
+      !config.authc.selector.enabled && anonymousProvider === config.authc.sortedProviders[0];
+
+    return {
+      isAnonymousAccessEnabled: this.isAnonymousAccessEnabled(),
+      accessURLParameters:
+        anonymousProvider && !anonymousIsDefault
+          ? Object.freeze(
+              new Map([[AUTH_PROVIDER_HINT_QUERY_STRING_PARAMETER, anonymousProvider.name]])
+            )
+          : null,
+      isSavedObjectTypeAccessibleAnonymously: async (request, savedObjectType) => {
+        this.logger.debug(`Checking if Saved Object is accessible anonymously.`);
+
+        // We should use credentials of the anonymous service account instead of credentials of the
+        // current user to figure out if the specified Saved Object type can be accessed anonymously.
+        const anonymousAuthorizationHeader = this.createAnonymousAuthorizationHeader();
+        const fakeAnonymousRequest = KibanaRequest.from(({
+          headers: anonymousAuthorizationHeader
+            ? { authorization: anonymousAuthorizationHeader.toString() }
+            : {},
+          path: '/',
+          route: { settings: {} },
+          url: { href: '/' },
+          raw: { req: { url: '/' } },
+        } as unknown) as Request);
+
+        const spaceId = spaces?.getSpaceId(request);
+        if (spaceId) {
+          basePath.set(fakeAnonymousRequest, addSpaceIdToPath('/', spaceId));
+        }
+
+        const checkPrivileges = authz.checkSavedObjectsPrivilegesWithRequest(fakeAnonymousRequest);
+        const { hasAllRequested } = await checkPrivileges(
+          authz.actions.savedObject.get(savedObjectType, 'get'),
+          spaces && spaceId ? spaces.spaceIdToNamespace(spaceId) : undefined
+        );
+
+        if (hasAllRequested) {
+          this.logger.debug(
```

Comment:
nit: While I'm generally a fan of logging all the things, these might be unnecessary. The return value here (which currently translates to the API route return value) isn't really transformed or fed into another server-side process.

Reply:
Was mostly using it to distinguish `canAccess: false` for the case when user cannot access something and when anonymous is not enabled. But yeah, it makes more sense to log it in the route instead (or wherever this API is used) :+1: